File: src/shell/bootSeed.ts
// First-paint boot seed (issue #814, ADR-0113): the Rust setup hook injects
// `window.__TOPTOPDUCK_BOOT_SEED__` via the main window's initialization
// script BEFORE any frontend script runs, so the persisted theme / locale
// paint immediately instead of flashing through the OS-resolved defaults
// while the full getAppConfig IPC is in flight. This module is the consuming
// boundary: parseBootSeed whitelists the two enum fields, readBootSeed is
// the thin reader useAppConfigState calls once per mount.
import { isLocalePreference } from "../i18n";
import type { LocalePreference, Theme } from "../types/app-config";

/** Keep in sync with the Rust writer (`src-tauri/src/app_config/boot_seed.rs`). */
export const BOOT_SEED_GLOBAL = "__TOPTOPDUCK_BOOT_SEED__";

export interface BootSeed {
  theme: Theme;
  locale: LocalePreference;
}

// The satisfies clause keeps the list in lockstep with the Theme union --
// a typo here would compile against a plain string list and silently
// reject every valid seed.
const THEMES: readonly string[] = [
  "system",
  "light",
  "dark",
] satisfies readonly Theme[];

function isTheme(value: unknown): value is Theme {
  return typeof value === "string" && THEMES.includes(value);
}

/** Whitelist the seed global: BOTH enum fields must validate or the whole
 * seed is dropped. Rust serializes typed enums, so an invalid variant can
 * only mean the global is not ours -- a dropped seed reproduces the
 * pre-#814 null behavior (the IPC full config stays the only authority).
 * Unknown extra keys are tolerated, mirroring the app-config serde read
 * (no deny_unknown_fields). */
export function parseBootSeed(raw: unknown): BootSeed | null {
  if (typeof raw !== "object" || raw === null) return null;
  const { theme, locale } = raw as Record<string, unknown>;
  if (!isTheme(theme)) return null;
  // The locale half reuses the IPC boundary guard so the seed whitelist and
  // the persisted-value whitelist can never drift apart (ADR-0113 Decision 3).
  if (!isLocalePreference(locale)) return null;
  return { theme, locale };
}

/** Read + validate the injected global. Null when the initialization script
 * did not run or its payload was dropped -- callers fall back. */
export function readBootSeed(): BootSeed | null {
  // Window carries no index signature for arbitrary injected globals, so the
  // read goes through unknown (the value is untrusted by definition -- the
  // whitelist in parseBootSeed is the trust boundary).
  const raw: unknown = (window as unknown as Record<string, unknown>)[BOOT_SEED_GLOBAL];
  return parseBootSeed(raw);
}

File: src-tauri/src/app_config/boot_seed.rs
//! First-paint boot seed (issue #814, ADR-0113): the narrow Rust -> webview
//! channel delivering `theme` + `locale` to the main window BEFORE any
//! frontend script runs, closing the first-paint flash where the persisted
//! preference differed from the OS-resolved default. `setup` attaches the
//! script at window creation (`create: false` + `from_config`); the payload
//! is intentionally TWO enum fields -- everything else waits for the
//! authoritative full `get_app_config` IPC. The honest-degrade read result
//! is injected unconditionally (defaults on a failed read are a semantic
//! no-op vs the pre-channel null fallback), so "the global is absent" means
//! exactly one thing: the script did not run.

use serde::Serialize;

use super::model::{AppConfig, LocalePreference, Theme};

/// The global the script assigns; read by `src/shell/bootSeed.ts`. The
/// double-underscore shape follows the tauri plugin convention
/// (`__TAURI_OS_PLUGIN_INTERNALS__`). Keep in sync with the frontend reader.
pub(crate) const BOOT_SEED_GLOBAL: &str = "__TOPTOPDUCK_BOOT_SEED__";

/// The wire payload. Reuses the model enums verbatim so the serde renames
/// keep the literals identical to the full `get_app_config` IPC wire by
/// construction (no parallel string mapping to drift).
#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct BootSeedPayload {
    pub theme: Theme,
    pub locale: LocalePreference,
}

/// Narrow the full config to the two first-paint fields.
pub(crate) fn boot_seed_payload(cfg: &AppConfig) -> BootSeedPayload {
    BootSeedPayload {
        theme: cfg.theme,
        locale: cfg.locale,
    }
}

/// The initialization script text. `serde_json` output over enum-only
/// fields is always a safe JSON literal (no free-form strings), so plain
/// `format!` carries no script-injection surface.
pub(crate) fn boot_seed_script(cfg: &AppConfig) -> String {
    let json = serde_json::to_string(&boot_seed_payload(cfg))
        .expect("serializing two unit enums to JSON cannot fail");
    format!("window.{BOOT_SEED_GLOBAL} = {json};")
}

#[cfg(test)]
mod tests {
    use super::*;

    fn config_with(theme: Theme, locale: LocalePreference) -> AppConfig {
        AppConfig {
            theme,
            locale,
            ..AppConfig::default()
        }
    }

    #[test]
    fn payload_narrows_to_the_two_first_paint_fields() {
        let payload = boot_seed_payload(&config_with(Theme::Dark, LocalePreference::ZhCN));
        assert_eq!(
            payload,
            BootSeedPayload {
                theme: Theme::Dark,
                locale: LocalePreference::ZhCN,
            }
        );
    }

    #[test]
    fn defaults_seed_system_system() {
        let script = boot_seed_script(&AppConfig::default());
        assert_eq!(
            script,
            "window.__TOPTOPDUCK_BOOT_SEED__ = {\"theme\":\"system\",\"locale\":\"system\"};"
        );
    }

    #[test]
    fn script_literals_match_the_ipc_wire_shapes() {
        // Lowercase theme (rename_all) + BCP-47 locale (explicit variant
        // renames): the exact literals the frontend whitelist accepts and
        // the full `get_app_config` IPC also emits.
        let script = boot_seed_script(&config_with(Theme::Dark, LocalePreference::ZhCN));
        assert_eq!(
            script,
            "window.__TOPTOPDUCK_BOOT_SEED__ = {\"theme\":\"dark\",\"locale\":\"zh-CN\"};"
        );
        let script = boot_seed_script(&config_with(Theme::Light, LocalePreference::EnUS));
        assert_eq!(
            script,
            "window.__TOPTOPDUCK_BOOT_SEED__ = {\"theme\":\"light\",\"locale\":\"en-US\"};"
        );
    }
}
